oeps/oep-0050-hooks-extension-framework.rst

```diff
@@ -0,0 +1,288 @@
+make the framework less complicated in its implementation.
+
+Also in the core platform code, the actual exposition of each particular hook
+could be configured (via flags or settings) so that even for big scales, at
+which the simple lookup for the functions or sending of the signals for a hook
+could have performance implications, is accounted for.
+
+Django 3.2 has just been released including the capabilities needed to use async
+patterns in the execution of views via coroutines. The usage of async await
+would be very beneficial for performance in both events and filters that are I/O
+heavy. The mechanism for calling filters should be implemented in a way that
+makes usage of this pattern possible. For the case of events, we expect Django
+to already implement their signals in the best possible way.
+
+Distributing the business logic into separate repos that are finally run in the
+same process, might be difficult to debug. The implementation of this proposal
+should include tools to track performance data as well as a glimpse into the
+current receivers of an event.
+
+
+Use Cases
+---------
+
+To further explain the sort of use cases that this proposal makes possible, we
+propose certain examples made easy by this framework.
+
+#. Generating a certificate in a third party system
+
+   The Event that fires after a certificate is generated in the platform is the
+   perfect place to call upon a service such as https://www.accredible.com/ or
+   https://www.sertifier.com to generate an external certificate. The same
+   applies to a badge in any badging system.
+
+#. Making enrollment possible for "premium" users in all courses
+
+   A Filter right before the enrollment could check the user metadata or call an
+   external system to determine if they are in a business tier that allows a
+   paying user to enroll in any course for an online academy. This use case
+   would require the hook called before the enrollment to support halting the
+   process as one of its responses to what the filter returns.
+
+#. Adding context information for an enterprise customer to a certificate
+
+   When rendering a certificate, a particular template that is stored in the
+   database may require certain information that is only available for users of
+   the particular course the certificate belongs to. This very custom logic can
+   be separated from the core platform repository via a Filter.
+
+
+Backward Compatibility
+======================
+
+The only current extension point that could be affected by this OEP is the
+REGISTRATION_EXTENSION_FORM. We suggest that this extension point continues to
+be supported until it is clear that migrating this to hooks is possible, and the
+migration path to the corresponding hook is well defined and documented.
+
+
+Reference Implementation
+========================
+
+The reference implementation must be completed before this OEP is given "Final"
+status. It is not necessary for the "accepted" status.
```

Comment:
[request] hmm... Can we replace "Final"  with "Accepted" and "accepted" with "Provisional"?

```
The reference implementation must be completed before this OEP is given "Accepted" status. It is not necessary for the "Provisional" status.
```

**Rationale:** We don't have any OEPs in the "Final" state and should probably update OEP-1 to exclude that state altogether.  "Accepted"  is synonymous with the community accepting and adopting the proposal.  We _did_ introduce a new state called "Provisional", which means the OEP was reviewed and tentatively accepted while the work is being implemented and piloted.